### PR TITLE
Recommend App: Implement share app button in About screen

### LINKED ISSUE #16997: Add sharing button in the About screen.
Parent issue: #16995

Implement the "Share WordPress with a friend" button in the About screen. This screen uses a plain table view, and a private `Row` structure to act as the cell's "view model".

### Designs

Screen Shot 2021-08-12 at 17 59 15

### Todo
- [x] Show the share app row based on feature flag.
- [x] Review accessibility.

### PR DESCRIPTION
**Pull request by dvdchr**
Refs #16995
Closes #16997 
Based on #17006

## Summary
This change adds the share app button to the About screen when the feature flag is enabled. While I did some clean up / tidying for the constants in `AboutViewController`, I tried to keep the implementation/flow as close as possible to the original since this is still feature flagged; once the flag is removed I can take a look at refactoring some parts of the class.

Here's a short video for the implementation:

https://user-images.githubusercontent.com/1299411/129255239-7559120f-9458-4624-b49d-83d13caa8ca6.mov

Notice that the after tapping the button for the first time, the button shows a loading indicator. However, in the later part of the video, the loading indicator no longer shown and the share sheet is displayed instantly. This is intentional, because the presenter is using a cached content and no longer needs to make another roundtrip.

## To Test

- Go to App Settings, and turn on the `Recommend App to Others` feature flag.
- Go to About screen.
- Verify that the share button is shown.
- Tap the share button, and verify that a loading indicator is shown before the share sheet is presented.
- Dismiss the share sheet.
- Tap the share button again, and verify that the loading indicator is no longer shown; the share sheet should be displayed immediately.

## Regression Notes
1. Potential unintended areas of impact
Some parts of the About screen is not displayed correctly. May impact self-hosted / non-WP accounts.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the screen, with and without the feature flag. Also tested with self-hosted accounts.

3. What automated tests I added (or what prevented me from doing so)
n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Me/App Settings/About/AboutViewController.swift [hunk 1]
```diff
@@ -217,49 +262,123 @@ open class AboutViewController: UITableViewController {
...
+        let account = AccountService(managedObjectContext: ContextManager.shared.mainContext).defaultWordPressComAccount()
+        let presenter = ShareAppContentPresenter(account: account)
+        presenter.delegate = self
+
+        return presenter
+    }()
+
+    /// Convenience property to determine whether the recomend app section should be displayed or not.
+    private var isRecommendAppSectionEnabled: Bool {
+        return FeatureFlag.recommendAppToOthers.enabled && !AppConfiguration.isJetpack
```
Comment (dvdchr) [inline WordPress/Classes/ViewRelated/Me/App Settings/About/AboutViewController.swift:282]: 🤔   @aerych @ScoutHarris Should we enable the share button feature for Jetpack app?

File: WordPress/Classes/ViewRelated/Me/App Settings/About/AboutViewController.swift [hunk 2]
```diff
@@ -174,33 +194,58 @@ open class AboutViewController: UITableViewController {
...
+    private func displayShareApp() {
+        tableView.deselectSelectedRowWithAnimation(true)
+
+        guard let sharePresenter = sharePresenter,
+              !sharePresenter.isLoading,
+              isRecommendAppSectionEnabled else {
+            return
+        }
+
+        sharePresenter.present(for: .wordpress, in: self)
```
Comment (dvdchr) [inline WordPress/Classes/ViewRelated/Me/App Settings/About/AboutViewController.swift]: Note to self: If we do decide to enable this feature for Jetpack, I'll need to feed the value based on `AppConfiguration.isJetpack`.

File: WordPress/Classes/ViewRelated/Me/App Settings/About/AboutViewController.swift (hunk 1 above)
Comment (aerych) [reply WordPress/Classes/ViewRelated/Me/App Settings/About/AboutViewController.swift:282]: Let's wait and release for WordPress first. Then if we need to iterate on anything we can do that and _then_ enable Jetpack.

**Review by aerych — COMMENTED** [review]
Heya @dvdchr 👋 
Tested and it's looking pretty good to me. I think it might be best to wait for any changes from https://github.com/wordpress-mobile/WordPress-iOS/pull/17005 and https://github.com/wordpress-mobile/WordPress-iOS/pull/17006 to be merged before finishing the review and testing. I'd like to do a complete test on the iPad, and I'm not seeing the button cell's selected state when I tap on the cell.  I _do_ see it shown in the example gif so I'm a bit puzzled about the difference.

File: WordPress/Classes/ViewRelated/Me/App Settings/About/AboutViewController.swift (hunk 1 above)
Comment (dvdchr) [reply WordPress/Classes/ViewRelated/Me/App Settings/About/AboutViewController.swift:282]: Sounds good! 👍

File: WordPress/Classes/ViewRelated/Me/App Settings/About/AboutViewController.swift (hunk 2 above)
Comment (dvdchr) [reply WordPress/Classes/ViewRelated/Me/App Settings/About/AboutViewController.swift]: https://github.com/wordpress-mobile/WordPress-iOS/pull/17007#discussion_r689818965

**Comment by dvdchr** [comment]
Hi @aerych. Thank you for your help! Referring to your comment https://github.com/wordpress-mobile/WordPress-iOS/pull/17006#pullrequestreview-733140482 on popover anchoring, I've addressed this in https://github.com/wordpress-mobile/WordPress-iOS/pull/17007/commits/e62da3bbeccb2f2c76d219e8d1072f8ebfee3cd6. 🙂 

> I'm not seeing the button cell's selected state when I tap on the cell.

This has also been addressed in this commit: https://github.com/wordpress-mobile/WordPress-iOS/commit/81dcbfcebf6b367495598a2f28f85436c25e135b which has been merged to `develop` and merged to this branch. So everything should now be ready for your second pass!

**Review by aerych — APPROVED** [review]
Looks great @dvdchr!  Tested with the updates and confirmed the highlighted cell and the popover looks good.  :shipit: !

**Comment by dvdchr** [comment]
Thanks @aerych 😄

**Pull request merged**